plugin/builtin/sortordertransformer/SortOrderTransformer.go

```diff
@@ -0,0 +1,134 @@
+	// if the appropriate fields are set in the kustomization.yaml file, which
+	// will also set the IsTopLevel field.
+	//
+	// See: https://github.com/kubernetes-sigs/kustomize/pull/4019#issuecomment-1159771417
+	IsTopLevel bool
+}
+
+//noinspection GoUnusedGlobalVariable
+var KustomizePlugin plugin
+
+// Nothing needed for configuration.
+func (p *plugin) Config(
+	_ *resmap.PluginHelpers, c []byte) (err error) {
+
+	// If config was called then this is not the root
+	err = yaml.Unmarshal(c, p)
+	if err != nil {
+		return
+	}
+
+	// Default to FIFO sort, aka no-op.
+	if p.SortOptions == nil {
+		p.SortOptions = &types.SortOptions{
+			Order: types.FIFOSortOrder,
+		}
+	}
+
+	// Validate that the only options set are the ones corresponding to the
+	// selected sort order.
+	if p.SortOptions.Order == types.FIFOSortOrder &&
+		p.SortOptions.LegacySortOptions != nil {
+		return errors.Errorf("the field 'sortOptions.legacySortOptions' is"+
+			" set but the selected sort order is '%v', not 'legacy'",
+			p.SortOptions.Order)
+	}
+
+	// If legacy sort is selected and no options are given, default to
+	// hardcoded order.
+	if p.SortOptions.Order == types.LegacySortOrder && p.SortOptions.LegacySortOptions == nil {
+		p.SortOptions.LegacySortOptions = &types.LegacySortOptions{
+			OrderFirst: resid.OrderFirst,
+			OrderLast:  resid.OrderLast,
+		}
+	}
+
+	return
+}
+
+func (p *plugin) Transform(m resmap.ResMap) (err error) {
+	if !p.IsTopLevel {
+		return nil
+	}
+	// Sort
+	resources := m.Resources()
+	if p.SortOptions.Order == types.LegacySortOrder {
+		s := &idSorter{resources: resources, sortOptions: p.SortOptions}
+		sort.Sort(s)
+		// Clear and refill with the correct order
+		m.Clear()
+		for _, r := range resources {
+			m.Append(r)
+		}
+	}
+
+	return nil
+}
+
+// idSorter implements the sort interface.
+type idSorter struct {
+	resources   []*resource.Resource
+	sortOptions *types.SortOptions
+}
+
+var _ sort.Interface = idSorter{}
+
+func (a idSorter) Len() int      { return len(a.resources) }
+func (a idSorter) Swap(i, j int) { a.resources[i], a.resources[j] = a.resources[j], a.resources[i] }
+func (a idSorter) Less(i, j int) bool {
+	if !a.resources[i].GetGvk().Equals(a.resources[j].GetGvk()) {
+		return gvkLessThan(a.resources[i].GetGvk(), a.resources[j].GetGvk(),
+			a.sortOptions.LegacySortOptions.OrderFirst,
+			a.sortOptions.LegacySortOptions.OrderLast)
+	}
+	return a.resources[i].CurId().String() < a.resources[j].CurId().String()
+}
+
+func gvkLessThan(gvk1, gvk2 resid.Gvk, orderFirst, orderLast []string) bool {
+	var typeOrders = func() map[string]int {
+		m := map[string]int{}
+		for i, n := range orderFirst {
+			m[n] = -len(orderFirst) + i
+		}
+		for i, n := range orderLast {
+			m[n] = 1 + i
+		}
+		return m
+	}()
+
+	index1 := typeOrders[gvk1.Kind]
+	index2 := typeOrders[gvk2.Kind]
+	if index1 != index2 {
+		return index1 < index2
+	}
+	return gvk1.String() < gvk2.String()
```

Comment:
I think this needs to be `legacySortString`--we had a regression in the past from using String directly, and I'm wondering how TestIssue4388 isn't catching this. 😕

Reply:
> I think this needs to be legacySortString--we had a regression in the past from using String directly, and I'm wondering how TestIssue4388 isn't catching this. 😕

Indeed, I also wonder about that 🤔 I will take a closer look.